### ISSUE TITLE
v0.7 errors

### ISSUE THREAD

**Issue by mohamed82008** [JuliaIntervals/IntervalArithmetic.jl#188]
So I opened this issue to discuss weird errors and causes in v0.7. The first one I find weird is:
```julia
julia> Interval(1) * (im)
ERROR: ArgumentError: cannot parse "false" as BigFloat
Stacktrace:
 [1] _parse_failure(::Type, ::String, ::Int64, ::Int64) at .\parse.jl:349 (repeats 2 times)
 [2] tryparse_internal at .\parse.jl:345 [inlined]
 [3] parse at .\parse.jl:355 [inlined]
 [4] #6 at C:\Users\user\.julia\dev\IntervalArithmetic\src\intervals\rounding.jl:58 [inlined]
 [5] setrounding(::getfield(IntervalArithmetic, Symbol("##6#8")){String}, ::Type{BigFloat}, ::RoundingMode{:Down}) at .\rounding.jl:158
 [6] #5 at C:\Users\user\.julia\dev\IntervalArithmetic\src\intervals\rounding.jl:57 [inlined]
 [7] setprecision(::getfield(IntervalArithmetic, Symbol("##5#7")){String,RoundingMode{:Down}}, ::Type{BigFloat}, ::Int64) at .\mpfr.jl:904
 [8] atomic at C:\Users\user\.julia\dev\IntervalArithmetic\src\intervals\rounding.jl:56 [inlined]
 [9] macro expansion at .\show.jl:556 [inlined]
 [10] convert(::Type{Interval{Float64}}, ::Bool) at C:\Users\user\.julia\dev\IntervalArithmetic\src\intervals\conversion.jl:16
 [11] _promote at .\promotion.jl:261 [inlined]
 [12] promote at .\promotion.jl:284 [inlined]
 [13] * at .\promotion.jl:314 [inlined]
 [14] *(::Interval{Float64}, ::Complex{Bool}) at .\complex.jl:299
 [15] top-level scope at none:0
```
This seems to be related to some promotion changes in Julia?

Using `1im` instead solves the problem.

**Comment by mohamed82008** [comment]
I think something is happening with the `convert` function. In v0.6.4, calling `Interval(1.)*im)` triggers the convert in the line 
https://github.com/JuliaIntervals/IntervalArithmetic.jl/blob/cab2433d6719899c38c0c3514fc935208614bdd0/src/intervals/conversion.jl#L16
with `x = 0`. While in v0.7, the same function is triggered with `x = false`. Calling the function with `x = false` in v0.6.4 triggers the same error. So it seems that Julia is doing something behind the scenes to convert the false to 0 in v0.6.4 but not anymore in v0.7.

**Comment by mohamed82008** [comment]
Stacktrace in v0.6.4:
```julia
 [1] convert at C:\Users\user\.julia\v0.6\IntervalArithmetic\src\intervals\conversion.jl:16 [inlined]
 [2] promote at .\promotion.jl:175 [inlined]
 [3] < at .\promotion.jl:291 [inlined]
 [4] signbit at .\number.jl:72 [inlined]
 [5] copysign(::IntervalArithmetic.Interval{Float64}, ::IntervalArithmetic.Interval{Float64}) at .\number.jl:110
 [6] *(::IntervalArithmetic.Interval{Float64}, ::Complex{Bool}) at .\complex.jl:254
```

Stacktrace in v0.7.0:
```julia
 [2] convert(::Type{Interval{Float64}}, ::Bool) at C:\Users\user\.julia\dev\IntervalArithmetic\src\intervals\conversion.jl:16
 [3] _promote at .\promotion.jl:261 [inlined]
 [4] promote at .\promotion.jl:284 [inlined]
 [5] *(::Interval{Float64}, ::Bool) at .\promotion.jl:314
 [6] *(::Interval{Float64}, ::Complex{Bool}) at .\complex.jl:299
 [7] top-level scope at none:0
```

**Comment by lbenet** [comment]
I cannot reproduce the error (on a mac), using Julia 0.7.0 and IntervalArithmetic v0.14.0:

```julia
julia> using IntervalArithmetic
[ Info: Recompiling stale cache file /Users/benet/.julia/compiled/v0.7/IntervalArithmetic/Gjmwo.ji for IntervalArithmetic [d1acc4aa-44c8-5952-acd4-ba5d80a2a253]
┌ Warning: `setrounding` for `Float32` and `Float64` has been deprecated, and will not be available in future versions.
│   caller = setrounding(::getfield(IntervalArithmetic, Symbol("##37#38")){Float64}, ::Type{Float64}, ::RoundingMode{:Down}) at rounding.jl:156
└ @ Base.Rounding ./rounding.jl:156
┌ Warning: `setrounding` for `Float32` and `Float64` has been deprecated, and will not be available in future versions.
│   caller = setrounding(::getfield(IntervalArithmetic, Symbol("##73#74")){Float64}, ::Type{Float64}, ::RoundingMode{:Up}) at rounding.jl:156
└ @ Base.Rounding ./rounding.jl:156

julia> Interval(1) * (im)
[0, 0] + [1, 1]*im

julia> Interval(1) * im
[0, 0] + [1, 1]*im

julia> 1im * Interval(1)
[0, 0] + [1, 1]*im

julia> im * Interval(1)
[0, 0] + [1, 1]*im

julia> versioninfo()
Julia Version 0.7.1-pre.0
Commit 36cddc1006 (2018-08-09 00:19 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin16.7.0)
  CPU: Intel(R) Core(TM) i7-6820HQ CPU @ 2.70GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.0 (ORCJIT, skylake)
Environment:
  JULIA_EDITOR = atom
```

**Comment by lbenet** [comment]
The only place where `iswindows()` is used, is to define two methods of `atomic`; these are in `src/intervals/conversion.jl`, [here](https://github.com/JuliaIntervals/IntervalArithmetic.jl/blob/cab2433d6719899c38c0c3514fc935208614bdd0/src/intervals/conversion.jl#L78) and [here](https://github.com/JuliaIntervals/IntervalArithmetic.jl/blob/cab2433d6719899c38c0c3514fc935208614bdd0/src/intervals/conversion.jl#L85). 

I guess that your code in windows somehow uses this methods. Can you check them directly?

**Comment by mohamed82008** [comment]
Yes it does. This function is the culprit.

https://github.com/JuliaIntervals/IntervalArithmetic.jl/blob/cab2433d6719899c38c0c3514fc935208614bdd0/src/intervals/rounding.jl#L55

```julia
julia> parse(Float64, string(false), RoundDown)
ERROR: ArgumentError: cannot parse "false" as BigFloat
Stacktrace:
 [1] _parse_failure(::Type, ::String, ::Int64, ::Int64) at .\parse.jl:349 (repeats 2 times)
 [2] tryparse_internal at .\parse.jl:345 [inlined]
 [3] parse at .\parse.jl:355 [inlined]
 [4] #6 at C:\Users\user\.julia\dev\IntervalArithmetic\src\intervals\rounding.jl:58 [inlined]
 [5] setrounding(::getfield(IntervalArithmetic, Symbol("##6#8")){String}, ::Type{BigFloat}, ::RoundingMode{:Down}) at .\rounding.jl:158
 [6] #5 at C:\Users\user\.julia\dev\IntervalArithmetic\src\intervals\rounding.jl:57 [inlined]
 [7] setprecision(::getfield(IntervalArithmetic, Symbol("##5#7")){String,RoundingMode{:Down}}, ::Type{BigFloat}, ::Int64) at .\mpfr.jl:904
 [8] parse(::Type{Float64}, ::String, ::RoundingMode{:Down}) at C:\Users\user\.julia\dev\IntervalArithmetic\src\intervals\rounding.jl:56
 [9] top-level scope at none:0
```

**Comment by mohamed82008** [comment]
In v0.6.4, `s` is "0", while in v0.7.0 it is "false", hence the error.

**Comment by mohamed82008** [comment]
Using
```julia
try
    parse(BigFloat, s)   # correctly takes account of rounding mode
catch
    BigFloat(parse(Bool, s))
end
```
solves the problem.

**Comment by mohamed82008** [comment]
PR?

**Comment by mohamed82008** [comment]
Another unusual warning I am not sure how to fix:
```julia
┌ Warning: `setrounding` for `Float32` and `Float64` has been deprecated, and will not be available in future versions.
│   caller = setrounding(::getfield(IntervalArithmetic, Symbol("##37#38")){Float64}, ::Type{Float64}, ::RoundingMode{:Down}) at rounding.jl:156
└ @ Base.Rounding .\rounding.jl:156
```
So what should it be replaced with?

**Comment by lbenet** [comment]
Regarding the PR, this is certainly welcome! So go ahead! Yet, I am not sure the `try`-`catch` is the best strategy, or at least, I do not understand how it solves it. In any case, please add the simple problematic case as a test.

What do you get with `@code_lowered`? I get this:
```julia
julia> @code_lowered Interval(1) * 1im
CodeInfo(
311 1 ─ %1 = (Base.real)(z)                                         │
    │   %2 = x * %1                                                 │
    │   %3 = (Base.imag)(z)                                         │
    │   %4 = x * %3                                                 │
    │   %5 = (Base.Complex)(%2, %4)                                 │
    └──      return %5                                              │
)
```

Any further comments on this, @dpsanders?

The weird warning, AFAICT, is unrelated, and is due to changes in 0.7, due to windows incompatibilities! We should certainly get rid of those before 1.0, but first things first.

**Comment by mohamed82008** [comment]
`1im` works but `im` doesn't. I get the same code lowered though:
```julia
julia> @code_lowered Interval(1) * im
CodeInfo(
299 1 ─ %1 = (Base.real)(z)                                                                                                                                                                                   │
    │   %2 = x * %1                                                                                                                                                                                           │
    │   %3 = (Base.imag)(z)                                                                                                                                                                                   │
    │   %4 = x * %3                                                                                                                                                                                           │
    │   %5 = (Base.Complex)(%2, %4)                                                                                                                                                                           │
    └──      return %5                                                                                                                                                                                        │
)
```

**Comment by mohamed82008** [comment]
`im` is `Complex{Bool}`, when it is passed to `real` or `imag`, it returns a `Bool`. The error happens when multiplying an interval by a boolean. In v0.6, the promotion mechanism saves us by somehow converting the boolean to an integer first. In v0.7, it doesn't so when it goes the through the Windows function, gets converted to a string becoming `"true"` or `"false"`, and then we try to parse it as "BigFloat", the error happens. So it is the interplay of the Windows function and the changes in promotion in Julia. Another perhaps more elegant fix would be to define how intervals multiply booleans as another method so it doesn't fallback on the `(::Inteveral, ::Real)` method.

**Comment by lbenet** [comment]
Thanks for looking at this! I agree, it is more elegant (julian) to add the new promotion rule, and let Julia do its job. Can you submit the PR?

**Comment by dpsanders** [comment]
Included in #201.
